### ISSUE TITLE
npm run dev constantly throws pm2 error on Windows Home machine

### ISSUE THREAD

**Issue by jakexconniff** [remix-run/blues-stack#31]
When I follow the instructions to setup the Blues Stack, `npm run dev` fails to ever start, and throws the error message which I have pasted below. It worked on my Macbook Pro, so I believe it is something related to Windows, potentially even specifically Windows Home.

OS: Windows 10 Home
OS build: 19043.1586
Version 21H1

Reproduction steps:

1. npx create-remix@latest
2.  Enter name-of-project
3. Select pre-existing stack option
4. Select Blues stack
5. Select Y to npm install
6. Wait for it to finish
7. cd name-of-project
8. npm run docker
9. Open Docker Desktop and confirm it has finished starting up.
10. npm run setup
11.  npm run build
12. npm run dev

```
$ npm run dev

> dev
> pm2-dev ./other/pm2.config.js

[
  Error: Script not found: C:\Users\myname\Code\blues-stack-test\node --inspect --require .\node_modules\dotenv\config --require .\mocks .\build\server.js
      at Object.Common.resolveAppAttributes (C:\Users\myname\Code\blues-stack-test\node_modules\pm2\lib\Common.js:647:11)
      at C:\Users\myname\Code\blues-stack-test\node_modules\pm2\lib\API.js:1129:35
      at C:\Users\myname\Code\blues-stack-test\node_modules\async\internal\withoutIndex.js:8:40
      at replenish (C:\Users\myname\Code\blues-stack-test\node_modules\async\internal\eachOfLimit.js:81:17)
      at C:\Users\myname\Code\blues-stack-test\node_modules\async\internal\eachOfLimit.js:86:9
      at eachLimit (C:\Users\myname\Code\blues-stack-test\node_modules\async\eachLimit.js:47:43)
      at awaitable (C:\Users\myname\Code\blues-stack-test\node_modules\async\internal\awaitify.js:13:28)
      at startApps (C:\Users\myname\Code\blues-stack-test\node_modules\pm2\lib\API.js:1113:7)
      at C:\Users\myname\Code\blues-stack-test\node_modules\pm2\lib\API.js:1094:16
      at wrapper (C:\Users\myname\Code\blues-stack-test\node_modules\async\internal\once.js:12:16),
  Error: Script not found: C:\Users\myname\Code\blues-stack-test\npm run build:server -- --watch
      at Object.Common.resolveAppAttributes (C:\Users\myname\Code\blues-stack-test\node_modules\pm2\lib\Common.js:647:11)
      at C:\Users\myname\Code\blues-stack-test\node_modules\pm2\lib\API.js:1129:35
      at C:\Users\myname\Code\blues-stack-test\node_modules\async\internal\withoutIndex.js:8:40
      at replenish (C:\Users\myname\Code\blues-stack-test\node_modules\async\internal\eachOfLimit.js:81:17)
      at C:\Users\myname\Code\blues-stack-test\node_modules\async\internal\eachOfLimit.js:86:9
      at eachLimit (C:\Users\myname\Code\blues-stack-test\node_modules\async\eachLimit.js:47:43)
      at awaitable (C:\Users\myname\Code\blues-stack-test\node_modules\async\internal\awaitify.js:13:28)
      at startApps (C:\Users\myname\Code\blues-stack-test\node_modules\pm2\lib\API.js:1113:7)
      at C:\Users\myname\Code\blues-stack-test\node_modules\pm2\lib\API.js:1094:16
      at wrapper (C:\Users\myname\Code\blues-stack-test\node_modules\async\internal\once.js:12:16),
  Error: Script not found: C:\Users\myname\Code\blues-stack-test\cross-env NODE_ENV=development remix watch
      at Object.Common.resolveAppAttributes (C:\Users\myname\Code\blues-stack-test\node_modules\pm2\lib\Common.js:647:11)
      at C:\Users\myname\Code\blues-stack-test\node_modules\pm2\lib\API.js:1129:35
      at C:\Users\myname\Code\blues-stack-test\node_modules\async\internal\withoutIndex.js:8:40
      at replenish (C:\Users\myname\Code\blues-stack-test\node_modules\async\internal\eachOfLimit.js:81:17)
      at C:\Users\myname\Code\blues-stack-test\node_modules\async\internal\eachOfLimit.js:86:9
      at eachLimit (C:\Users\myname\Code\blues-stack-test\node_modules\async\eachLimit.js:47:43)
      at awaitable (C:\Users\myname\Code\blues-stack-test\node_modules\async\internal\awaitify.js:13:28)
      at startApps (C:\Users\myname\Code\blues-stack-test\node_modules\pm2\lib\API.js:1113:7)
      at C:\Users\myname\Code\blues-stack-test\node_modules\pm2\lib\API.js:1094:16
      at wrapper (C:\Users\myname\Code\blues-stack-test\node_modules\async\internal\once.js:12:16),
  Error: Script not found: C:\Users\myname\Code\blues-stack-test\npm run generate:css -- --watch
      at Object.Common.resolveAppAttributes (C:\Users\myname\Code\blues-stack-test\node_modules\pm2\lib\Common.js:647:11)
      at C:\Users\myname\Code\blues-stack-test\node_modules\pm2\lib\API.js:1129:35
      at C:\Users\myname\Code\blues-stack-test\node_modules\async\internal\withoutIndex.js:8:40
      at replenish (C:\Users\myname\Code\blues-stack-test\node_modules\async\internal\eachOfLimit.js:81:17)
      at C:\Users\myname\Code\blues-stack-test\node_modules\async\internal\eachOfLimit.js:86:9
      at eachLimit (C:\Users\myname\Code\blues-stack-test\node_modules\async\eachLimit.js:47:43)
      at awaitable (C:\Users\myname\Code\blues-stack-test\node_modules\async\internal\awaitify.js:13:28)
      at startApps (C:\Users\myname\Code\blues-stack-test\node_modules\pm2\lib\API.js:1113:7)
      at C:\Users\myname\Code\blues-stack-test\node_modules\pm2\lib\API.js:1094:16
      at wrapper (C:\Users\myname\Code\blues-stack-test\node_modules\async\internal\once.js:12:16)
]
```

**Comment by jakexconniff** [comment]
Yeah it looks like all 4 scripts are working in separate terminals. In order, here is the output from the terminals in case it is ever relevant:

```
$ node --inspect --require .\node_modules\dotenv\config --require .\mocks .\build\server.js
Debugger listening on ws://127.0.0.1:9229/dadba26d-186a-4f72-a4c3-2ebfc3f049df
For help, see: https://nodejs.org/en/docs/inspector
� Mock server running
(node:19060) ExperimentalWarning: stream/web is an experimental feature. This feature could change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
� setting up prisma client to localhost:5432
✅ app ready: http://localhost:3000
```

```
$ npm run build:server -- --watch

> build:server
> esbuild --platform=node --format=cjs ./server.ts --outdir=build "--watch"

[watch] build finished, watching for changes...
```

```
$ npx remix watch
Watching Remix app in development mode...
� Built in 693ms
� File changed: app\styles\tailwind.css
� Rebuilding...
� Rebuilt in 227ms
```

```
$ npm run generate:css -- --watch

> generate:css
> tailwindcss -o ./app/styles/tailwind.css "--watch"


Rebuilding...
Done in 286ms.
```

**Comment by kentcdodds** [comment]
Yeah, so this looks like just a matter of pm2 doing something weird with the node script, but it appears to be fine with npm scripts, so maybe try the latest push I just made where I moved it into an npm script: https://github.com/remix-run/blues-stack/commit/89bb536e37ac1ff4fca82152d8fd97c3546b35c6

**Comment by OriginalEXE** [comment]
Hi,
 
I ran into the same issue, it's pm2 related, happens only on Windows and there are no non-hacky solutions it seems. You can read more about the issue here: https://github.com/Unitech/pm2/issues/3657

TL;DR; For some weird reason, pm2 on Windows assumes that whatever you pass to `script` inside of the `apps` array of your `pm2.config.json` is a path to a file, rather than a command (even though it properly supports that same command when you run it as a single command). So my approach is to actually create files for each of those commands and then use `shelljs` package to spawn shells which run the actual commands.

I have fixed it by doing the following:

1. Change `/other/pm2.config.js` to:

```js
const env = {
  NODE_ENV: process.env.NODE_ENV ?? "development",
  FORCE_COLOR: "1",
};

module.exports = {
  apps: [
    {
      name: "Server",
      script: "./other/server.js",
      watch: ["./mocks/**/*.ts", "./build/server.js", "./.env"],
      env,
    },
    {
      name: "Server Build",
      script: "./other/server-build.js",
      ignore_watch: ["."],
      env,
    },
    {
      name: "Remix",
      script: "./other/remix.js",
      ignore_watch: ["."],
      env,
    },
    {
      name: "Tailwind",
      script: "./other/tailwind.js",
      autorestart: false,
      ignore_watch: ["."],
      env,
    },
  ],
};
```

2. Create `/other/server.js` with the following:
```js
var shell = require("shelljs");

shell.exec("npm run dev:server", { windowsHide: true });
```

3. Create `/other/server-build.js` with the following:
```js
var shell = require("shelljs");

shell.exec("npm run build:server -- --watch", { windowsHide: true });
```

4. Create `/other/remix.js` with the following:
```js
var shell = require("shelljs");

shell.exec("npx remix watch", { windowsHide: true });
```

5. Create `/other/tailwind.js` wit the following:
```js
var shell = require("shelljs");

shell.exec("npm run generate:css -- --watch", { windowsHide: true });
```

Your `npm run dev` should now work as expected.
